src/execution_plan/optimizations/reduce_traversal.c

```diff
@@ -0,0 +1,75 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "reduce_traversal.h"
+
+#include "../../util/arr.h"
+#include "../../util/vector.h"
+#include "../ops/op_expand_into.h"
+#include "../ops/op_conditional_traverse.h"
+
+static bool _entity_resolved(OpBase *root, const char *entity) {
+    int count = (root->modifies) ? Vector_Size(root->modifies) : 0;
+
+    for(int i = 0; i < count; i++) {
+        char *alias;
+        Vector_Get(root->modifies, i, &alias);
+        if(strcmp(entity, alias) == 0) return true;
+    }
+
+    for(int i = 0; i < root->childCount; i++) {
+        OpBase *child = root->children[i];
+        if(_entity_resolved(child, entity)) return true;
+    }
+
+    return false;
+}
+
+/* Inspect each traverse operation T,
+ * For each T see if T's source and destination nodes
+ * are already resolved, in which case replace traversal operation 
+ * with expand-into op. */
+void reduceTraversal(ExecutionPlan *plan) {
+    // OPType t = OPType_CONDITIONAL_TRAVERSE | OPType_CONDITIONAL_VAR_LEN_TRAVERSE;
+    OPType t = OPType_CONDITIONAL_TRAVERSE;
+    OpBase **traversals = ExecutionPlan_LocateOps(plan->root, t);
+    uint traversals_count = array_len(traversals);
+    for(uint i = 0; i < traversals_count; i++) {
+        CondTraverse *traverse = (CondTraverse*)traversals[i];
+
+        /* If traverse src and dest nodes are the same,
+         * traverse acts as a filter which make sure the node is of a specific type
+         * e.g. MATCH (a:A)-[e*]->(b:B) RETURN e 
```

Comment:
Aliases on variable-length edges is a hazy issue in the Cypher specification that I expect will change a lot in upcoming versions. In any case, our grammar rules don't support this query, nor can we support it right now, since returning `e` must take the form of a path if the number of hops is > 1.

Reply:
👍